stdlib/numbers.pyi

```diff
@@ -1,130 +1,154 @@
+# (since type checkers don't see `complex` as a subtype of `numbers.Complex`,
+# nor `float` as a subtype of `numbers.Real`, etc.)
 
+import sys
 from _typeshed import Incomplete
 from abc import ABCMeta, abstractmethod
-from typing import SupportsFloat, overload
+from typing import Literal, SupportsFloat, SupportsIndex, overload
+from typing_extensions import TypeAlias
+
+if sys.version_info >= (3, 11):
+    from typing import SupportsComplex as _SupportsComplex
+else:
+    # builtins.complex didn't have a __complex__ method on older Pythons
+    import typing
+
+    _SupportsComplex: TypeAlias = typing.SupportsComplex | complex
 
 __all__ = ["Number", "Complex", "Real", "Rational", "Integral"]
 
 class Number(metaclass=ABCMeta):
     @abstractmethod
     def __hash__(self) -> int: ...
 
+# See comment at the top of the file
+# for why some of these return types are purposefully vague
 class Complex(Number):
     @abstractmethod
     def __complex__(self) -> complex: ...
     def __bool__(self) -> bool: ...
     @property
     @abstractmethod
-    def real(self): ...
+    def real(self) -> SupportsFloat: ...
     @property
     @abstractmethod
-    def imag(self): ...
+    def imag(self) -> SupportsFloat: ...
     @abstractmethod
-    def __add__(self, other): ...
+    def __add__(self, other) -> _SupportsComplex: ...
     @abstractmethod
-    def __radd__(self, other): ...
+    def __radd__(self, other) -> _SupportsComplex: ...
     @abstractmethod
-    def __neg__(self): ...
+    def __neg__(self) -> _SupportsComplex: ...
     @abstractmethod
-    def __pos__(self): ...
-    def __sub__(self, other): ...
-    def __rsub__(self, other): ...
+    def __pos__(self) -> _SupportsComplex: ...
+    def __sub__(self, other) -> _SupportsComplex: ...
+    def __rsub__(self, other) -> _SupportsComplex: ...
     @abstractmethod
-    def __mul__(self, other): ...
+    def __mul__(self, other) -> _SupportsComplex: ...
     @abstractmethod
-    def __rmul__(self, other): ...
+    def __rmul__(self, other) -> _SupportsComplex: ...
     @abstractmethod
-    def __truediv__(self, other): ...
+    def __truediv__(self, other) -> _SupportsComplex: ...
     @abstractmethod
-    def __rtruediv__(self, other): ...
+    def __rtruediv__(self, other) -> _SupportsComplex: ...
     @abstractmethod
-    def __pow__(self, exponent): ...
+    def __pow__(self, exponent) -> _SupportsComplex: ...
     @abstractmethod
-    def __rpow__(self, base): ...
+    def __rpow__(self, base) -> _SupportsComplex: ...
     @abstractmethod
-    def __abs__(self) -> Real: ...
+    def __abs__(self) -> SupportsFloat: ...
     @abstractmethod
-    def conjugate(self): ...
+    def conjugate(self) -> _SupportsComplex: ...
     @abstractmethod
     def __eq__(self, other: object) -> bool: ...
 
+# See comment at the top of the file
+# for why some of these return types are purposefully vague
 class Real(Complex, SupportsFloat):
```

Comment:
I think Real needs to override some methods from Complex to say they support SupportsIndex rather than _SupportsComplex. For example, saying that `__sub__` should return SupportsFloat rather than SupportsComplex may make this error in ibis go away: https://github.com/ibis-project/ibis/blob/0db47322595f93065997c990920b3b89475a1f9e/ibis/util.py#L168

Reply:
Unfortunately that wouldn't be sound, since it's perfectly legal to subtract a `Complex` from a ` Real`, and the result won't necessarily be a `Real`:

```pycon
>>> isinstance(1.2, numbers.Real)
True
>>> isinstance(1.2 - 1j, numbers.Real)
False
```

Reply:
Good point. Maybe an overload that takes a Real argument and returns SupportsFloat would help, but that sounds like a rabbit hole that might cause more problems.

Reply:
Yeah, definitely not something for this PR